kolena/_experimental/workflow/thresholded.py

```diff
@@ -0,0 +1,121 @@
+# Copyright 2021-2023 Kolena Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from abc import ABCMeta
+from dataclasses import dataclass
+from dataclasses import fields
+
+from kolena.workflow._datatypes import _register_data_type
+from kolena.workflow._datatypes import DataType
+from kolena.workflow._datatypes import TypedDataObject
+
+
+class PreventThresholdOverrideMeta(ABCMeta, type):
+    def __new__(cls, name, bases, dct):
+        if "threshold" in dct.get("__annotations__", {}):
+            for base in bases:
+                if base.__name__ == "ThresholdedMetrics":
+                    raise TypeError(f"Subclasses of {base.__name__} cannot override 'threshold'")
+        return super().__new__(cls, name, bases, dct)
+
+
+class _MetricsType(DataType):
+    THRESHOLDED = "THRESHOLDED"
+
+    @staticmethod
+    def _data_category() -> str:
+        return "METRICS"
+
+
+@dataclass(frozen=True)
+class ThresholdedMetrics(TypedDataObject[_MetricsType], metaclass=PreventThresholdOverrideMeta):
+    """
+    Represents metrics tied to a specific threshold.
```

Comment:
Does this block show up in the docs when you render things locally? You might need to update `docs/reference/workflow/evaluator.md`